--- plymouth-0.9.2-old/src/libply-splash-core/ply-device-manager.c  2017-11-15 10:14:15.465576724 -0700
+++ plymouth-0.9.2/src/libply-splash-core/ply-device-manager.c  2017-11-15 10:14:47.301543585 -0700
@@ -106,7 +106,7 @@
         boot_vga = udev_device_get_sysattr_value (bus_device, "boot_vga");
         free (bus_device_path);
 
-        if (boot_vga != NULL && strcmp (boot_vga, "1") == 0)
+        if (boot_vga == NULL || strcmp (boot_vga, "1") == 0)
                 for_local_console = true;
         else
                 for_local_console = false;

