diff --recursive --unified --new-file a/src/admin_rights.cpp b/src/admin_rights.cpp
--- a/src/admin_rights.cpp	2012-10-31 23:50:46.000000000 -0600
+++ b/src/admin_rights.cpp	2013-06-17 22:40:43.251309748 -0600
@@ -64,7 +64,9 @@
     m_lock(m_mutex), m_user_id(-1), m_has_rights(false), m_use_log(use_log)
 {
     m_user_id = geteuid();
-    if ( seteuid(ADMIN_USER_ID) != 0 ) {
+	if ( m_user_id == ADMIN_USER_ID ) {
+		return;
+	} else if ( seteuid(ADMIN_USER_ID) != 0 ) {
         if (m_use_log)
             PION_LOG_ERROR(m_logger, "Unable to upgrade to administrative rights");
         m_lock.unlock();
