diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/client/StaticHostProvider.java b/src/java/main/org/apache/zookeeper/client/StaticHostProvider.java
--- a/src/java/main/org/apache/zookeeper/client/StaticHostProvider.java	2013-10-22 21:06:53.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/client/StaticHostProvider.java	2013-10-22 21:25:47.162573812 -0600
@@ -114,7 +114,7 @@
         for (InetSocketAddress address : serverAddresses) {
             InetAddress ia = address.getAddress();
             InetAddress resolvedAddresses[] = InetAddress.getAllByName((ia!=null) ? ia.getHostAddress():
-                    address.getHostName());
+                    address.getHostString());
             for (InetAddress resolvedAddress : resolvedAddresses) {
                 tmpList.add(new InetSocketAddress(resolvedAddress
                         .getHostAddress(), address.getPort()));
@@ -161,7 +161,7 @@
             if (addr.getPort() == currentHost.getPort() &&
                     ((addr.getAddress()!=null && currentHost.getAddress()!=null &&
                       addr.getAddress().equals(currentHost.getAddress()))
-                     || addr.getHostName().equals(currentHost.getHostName()))) {
+                     || addr.toString().equals(currentHost.toString()))) {
                    myServerInNewConfig = true;
                    break;
                }
diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/ClientCnxn.java b/src/java/main/org/apache/zookeeper/ClientCnxn.java
--- a/src/java/main/org/apache/zookeeper/ClientCnxn.java	2013-10-22 21:06:54.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/ClientCnxn.java	2013-10-22 21:25:47.162573812 -0600
@@ -978,10 +978,10 @@
             }
 
             setName(getName().replaceAll("\\(.*\\)",
-                    "(" + addr.getHostName() + ":" + addr.getPort() + ")"));
+                    "(" + addr.getHostString() + ":" + addr.getPort() + ")"));
             if (ZooKeeperSaslClient.isEnabled()) {
                 try {
-                    zooKeeperSaslClient = new ZooKeeperSaslClient("zookeeper/"+addr.getHostName());
+                    zooKeeperSaslClient = new ZooKeeperSaslClient("zookeeper/"+addr.getHostString());
                 } catch (LoginException e) {
                     // An authentication error occurred when the SASL client tried to initialize:
                     // for Kerberos this means that the client failed to authenticate with the KDC.
@@ -1164,7 +1164,7 @@
             Socket sock = null;
             BufferedReader br = null;
             try {
-                sock = new Socket(addr.getHostName(), addr.getPort());
+                sock = new Socket(addr.getHostString(), addr.getPort());
                 sock.setSoLinger(false, -1);
                 sock.setSoTimeout(1000);
                 sock.setTcpNoDelay(true);
@@ -1202,8 +1202,7 @@
                 // save the found address so that it's used during the next
                 // connection attempt
                 rwServerAddress = addr;
-                throw new RWServerFoundException("Majority server found at "
-                        + addr.getHostName() + ":" + addr.getPort());
+                throw new RWServerFoundException("Majority server found at " + addr);
             }
         }
 
diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java b/src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java
--- a/src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java	2013-10-22 21:06:54.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java	2013-10-22 21:25:47.166573816 -0600
@@ -179,7 +179,7 @@
             // represents protocol version (in other words - message type)
             dout.writeLong(0xffff0000);              
             dout.writeLong(self.getId());
-            String addr = self.getElectionAddress().getHostName() + ":" + self.getElectionAddress().getPort();
+            String addr = self.getElectionAddress().getHostString() + ":" + self.getElectionAddress().getPort();
             byte[] addr_bytes = addr.getBytes();
             dout.writeInt(addr_bytes.length);
             dout.write(addr_bytes);
diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java b/src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java
--- a/src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java	2013-10-22 21:06:54.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java	2013-10-22 21:25:47.166573816 -0600
@@ -220,7 +220,7 @@
             StringWriter sw = new StringWriter();            
             //addr should never be null, but just to make sure
             if (addr !=null) { 
-                sw.append(addr.getHostName());
+                sw.append(addr.getHostString());
                 sw.append(":");
                 sw.append(String.valueOf(addr.getPort()));
             }
@@ -232,7 +232,7 @@
             else if (type == LearnerType.PARTICIPANT) sw.append(":participant");            
             if (clientAddr!=null){
                 sw.append(";");
-                sw.append(clientAddr.getHostName() + ":" + String.valueOf(clientAddr.getPort()));
+                sw.append(clientAddr.getHostString() + ":" + String.valueOf(clientAddr.getPort()));
             }
             return sw.toString();       
         }
diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/server/quorum/RemotePeerBean.java b/src/java/main/org/apache/zookeeper/server/quorum/RemotePeerBean.java
--- a/src/java/main/org/apache/zookeeper/server/quorum/RemotePeerBean.java	2013-10-22 21:06:54.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/server/quorum/RemotePeerBean.java	2013-10-22 21:25:47.166573816 -0600
@@ -39,7 +39,7 @@
     }
 
     public String getQuorumAddress() {
-        return peer.addr.getHostName()+":"+peer.addr.getPort();
+        return peer.addr.getHostString()+":"+peer.addr.getPort();
     }
 
 }
diff --recursive --unified --new-file a/src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java b/src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java
--- a/src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java	2013-10-22 21:06:54.000000000 -0600
+++ b/src/java/main/org/apache/zookeeper/server/util/ConfigUtils.java	2013-10-22 21:25:47.166573816 -0600
@@ -53,7 +53,7 @@
              }
              if (!first) sb.append(",");
              else first = false;
-             sb.append(qs.clientAddr.getHostName() + ":" + qs.clientAddr.getPort());
+             sb.append(qs.clientAddr.getHostString() + ":" + qs.clientAddr.getPort());
         }
         return version + " " + sb.toString();
     }
