Index: /trunk/boost/asio/ssl/detail/stream_core.hpp
===================================================================
--- a/boost/asio/ssl/detail/stream_core.hpp	(revision 70096)
+++ b/boost/asio/ssl/detail/stream_core.hpp	(revision 74817)
@@ -35,11 +35,15 @@
 struct stream_core
 {
+  // According to the OpenSSL documentation, this is the buffer size that is is
+  // sufficient to hold the largest possible TLS record.
+  enum { max_tls_record_size = 17 * 1024 };
+
   stream_core(SSL_CTX* context, boost::asio::io_service& io_service)
     : engine_(context),
       pending_read_(io_service),
       pending_write_(io_service),
-      output_buffer_space_(16384),
+      output_buffer_space_(max_tls_record_size),
       output_buffer_(boost::asio::buffer(output_buffer_space_)),
-      input_buffer_space_(16384),
+      input_buffer_space_(max_tls_record_size),
       input_buffer_(boost::asio::buffer(input_buffer_space_))
   {
