diff -r 81fd62cb4605 -r 5047b8f60327 boost/boost/config.hpp
--- a/boost/config.hpp	Wed Mar 23 15:07:03 2011 -0400
+++ b/boost/config.hpp	Thu Mar 24 13:10:49 2011 -0400
@@ -17,6 +17,8 @@
 #ifndef BOOST_CONFIG_HPP
 #define BOOST_CONFIG_HPP
 
+#define BOOST_FILESYSTEM_VERSION 2
+
 // if we don't have a user config, then use the default location:
 #if !defined(BOOST_USER_CONFIG) && !defined(BOOST_NO_USER_CONFIG)
 #  define BOOST_USER_CONFIG <boost/config/user.hpp>
diff -r 81fd62cb4605 -r 5047b8f60327 boost/boost/filesystem/config.hpp
--- a/boost/filesystem/config.hpp	Wed Mar 23 15:07:03 2011 -0400
+++ b/boost/filesystem/config.hpp	Thu Mar 24 13:10:49 2011 -0400
@@ -21,7 +21,7 @@
 # endif
 
 # if !defined(BOOST_FILESYSTEM_VERSION)
-#   define BOOST_FILESYSTEM_VERSION 3
+#   define BOOST_FILESYSTEM_VERSION 2
 # endif
 
 #if BOOST_FILESYSTEM_VERSION == 2
