diff --recursive --unified --new-file a/src/lib_json/json_writer.cpp b/src/lib_json/json_writer.cpp
--- a/src/lib_json/json_writer.cpp	2011-05-02 15:47:24.000000000 -0600
+++ b/src/lib_json/json_writer.cpp	2013-06-17 22:40:53.051327960 -0600
@@ -208,7 +208,6 @@
 {
    document_ = "";
    writeValue( root );
-   document_ += "\n";
    return document_;
 }
 
@@ -829,8 +828,8 @@
 
 std::ostream& operator<<( std::ostream &sout, const Value &root )
 {
-   Json::StyledStreamWriter writer;
-   writer.write(sout, root);
+   Json::FastWriter writer;
+   sout << writer.write(root);
    return sout;
 }
 
