diff --recursive --unified --new-file a/Makefile.in b/Makefile.in
--- a/Makefile.in	2012-02-22 14:09:17.000000000 -0700
+++ b/Makefile.in	2013-06-17 22:41:19.131376426 -0600
@@ -41,7 +41,7 @@
 	libc_allocator_with_realloc_test$(EXEEXT) \
 	sparsetable_unittest$(EXEEXT) hashtable_test$(EXEEXT) \
 	simple_test$(EXEEXT) simple_compat_test$(EXEEXT)
-noinst_PROGRAMS = $(am__EXEEXT_1) time_hash_map$(EXEEXT)
+noinst_PROGRAMS =
 subdir = .
 DIST_COMMON = README $(am__configure_deps) $(dist_doc_DATA) \
 	$(googleinclude_HEADERS) $(googleinternalinclude_HEADERS) \
