diff -ur gmock-1.6.0.orig/gtest/scripts/fuse_gtest_files.py gmock-1.6.0/gtest/scripts/fuse_gtest_files.py
--- gmock-1.6.0.orig/gtest/scripts/fuse_gtest_files.py	2010-11-03 15:14:13.000000000 -0600
+++ gmock-1.6.0/gtest/scripts/fuse_gtest_files.py	2016-09-27 16:01:43.240121501 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/scripts/gen_gtest_pred_impl.py gmock-1.6.0/gtest/scripts/gen_gtest_pred_impl.py
--- gmock-1.6.0.orig/gtest/scripts/gen_gtest_pred_impl.py	2010-11-03 15:14:13.000000000 -0600
+++ gmock-1.6.0/gtest/scripts/gen_gtest_pred_impl.py	2016-09-27 16:01:45.936147225 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/scripts/pump.py gmock-1.6.0/gtest/scripts/pump.py
--- gmock-1.6.0.orig/gtest/scripts/pump.py	2010-12-02 16:29:19.000000000 -0700
+++ gmock-1.6.0/gtest/scripts/pump.py	2016-09-27 16:01:41.416104065 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_break_on_failure_unittest.py gmock-1.6.0/gtest/test/gtest_break_on_failure_unittest.py
--- gmock-1.6.0.orig/gtest/test/gtest_break_on_failure_unittest.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_break_on_failure_unittest.py	2016-09-27 16:01:56.551248647 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_catch_exceptions_test.py gmock-1.6.0/gtest/test/gtest_catch_exceptions_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_catch_exceptions_test.py	2010-12-02 16:29:18.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_catch_exceptions_test.py	2016-09-27 16:01:51.375199189 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2010 Google Inc.  All Rights Reserved.
 #
diff -ur gmock-1.6.0.orig/gtest/test/gtest_color_test.py gmock-1.6.0/gtest/test/gtest_color_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_color_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_color_test.py	2016-09-27 16:02:13.295408811 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_env_var_test.py gmock-1.6.0/gtest/test/gtest_env_var_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_env_var_test.py	2010-12-02 16:29:18.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_env_var_test.py	2016-09-27 16:01:47.855165551 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_filter_unittest.py gmock-1.6.0/gtest/test/gtest_filter_unittest.py
--- gmock-1.6.0.orig/gtest/test/gtest_filter_unittest.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_filter_unittest.py	2016-09-27 16:01:49.624182482 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2005 Google Inc. All Rights Reserved.
 #
diff -ur gmock-1.6.0.orig/gtest/test/gtest_help_test.py gmock-1.6.0/gtest/test/gtest_help_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_help_test.py	2010-12-02 16:29:18.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_help_test.py	2016-09-27 16:02:09.911376429 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_list_tests_unittest.py gmock-1.6.0/gtest/test/gtest_list_tests_unittest.py
--- gmock-1.6.0.orig/gtest/test/gtest_list_tests_unittest.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_list_tests_unittest.py	2016-09-27 16:02:08.167359760 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_output_test.py gmock-1.6.0/gtest/test/gtest_output_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_output_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_output_test.py	2016-09-27 16:01:58.399266321 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_shuffle_test.py gmock-1.6.0/gtest/test/gtest_shuffle_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_shuffle_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_shuffle_test.py	2016-09-27 16:02:00.239283921 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009 Google Inc. All Rights Reserved.
 #
diff -ur gmock-1.6.0.orig/gtest/test/gtest_test_utils.py gmock-1.6.0/gtest/test/gtest_test_utils.py
--- gmock-1.6.0.orig/gtest/test/gtest_test_utils.py	2011-02-23 12:41:23.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_test_utils.py	2016-09-27 16:02:15.135426423 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_throw_on_failure_test.py gmock-1.6.0/gtest/test/gtest_throw_on_failure_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_throw_on_failure_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_throw_on_failure_test.py	2016-09-27 16:02:11.743393947 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_uninitialized_test.py gmock-1.6.0/gtest/test/gtest_uninitialized_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_uninitialized_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_uninitialized_test.py	2016-09-27 16:02:06.600344755 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_xml_outfiles_test.py gmock-1.6.0/gtest/test/gtest_xml_outfiles_test.py
--- gmock-1.6.0.orig/gtest/test/gtest_xml_outfiles_test.py	2010-11-03 15:14:10.000000000 -0600
+++ gmock-1.6.0/gtest/test/gtest_xml_outfiles_test.py	2016-09-27 16:02:02.031301057 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_xml_output_unittest.py gmock-1.6.0/gtest/test/gtest_xml_output_unittest.py
--- gmock-1.6.0.orig/gtest/test/gtest_xml_output_unittest.py	2011-02-01 18:26:23.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_xml_output_unittest.py	2016-09-27 16:01:52.991214642 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/test/gtest_xml_test_utils.py gmock-1.6.0/gtest/test/gtest_xml_test_utils.py
--- gmock-1.6.0.orig/gtest/test/gtest_xml_test_utils.py	2011-02-01 18:26:23.000000000 -0700
+++ gmock-1.6.0/gtest/test/gtest_xml_test_utils.py	2016-09-27 16:01:54.615230165 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/gtest/xcode/Scripts/versiongenerate.py gmock-1.6.0/gtest/xcode/Scripts/versiongenerate.py
--- gmock-1.6.0.orig/gtest/xcode/Scripts/versiongenerate.py	2010-11-03 15:14:16.000000000 -0600
+++ gmock-1.6.0/gtest/xcode/Scripts/versiongenerate.py	2016-09-27 16:02:17.119445417 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/scripts/fuse_gmock_files.py gmock-1.6.0/scripts/fuse_gmock_files.py
--- gmock-1.6.0.orig/scripts/fuse_gmock_files.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/fuse_gmock_files.py	2016-09-27 16:01:39.480085584 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/scripts/generator/cpp/ast.py gmock-1.6.0/scripts/generator/cpp/ast.py
--- gmock-1.6.0.orig/scripts/generator/cpp/ast.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/cpp/ast.py	2016-09-27 16:01:37.776069325 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2007 Neal Norwitz
 # Portions Copyright 2007 Google Inc.
diff -ur gmock-1.6.0.orig/scripts/generator/cpp/gmock_class.py gmock-1.6.0/scripts/generator/cpp/gmock_class.py
--- gmock-1.6.0.orig/scripts/generator/cpp/gmock_class.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/cpp/gmock_class.py	2016-09-27 16:01:30.528000126 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008 Google Inc.  All Rights Reserved.
 #
diff -ur gmock-1.6.0.orig/scripts/generator/cpp/keywords.py gmock-1.6.0/scripts/generator/cpp/keywords.py
--- gmock-1.6.0.orig/scripts/generator/cpp/keywords.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/cpp/keywords.py	2016-09-27 16:01:32.184015970 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2007 Neal Norwitz
 # Portions Copyright 2007 Google Inc.
diff -ur gmock-1.6.0.orig/scripts/generator/cpp/tokenize.py gmock-1.6.0/scripts/generator/cpp/tokenize.py
--- gmock-1.6.0.orig/scripts/generator/cpp/tokenize.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/cpp/tokenize.py	2016-09-27 16:01:33.816031541 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2007 Neal Norwitz
 # Portions Copyright 2007 Google Inc.
diff -ur gmock-1.6.0.orig/scripts/generator/cpp/utils.py gmock-1.6.0/scripts/generator/cpp/utils.py
--- gmock-1.6.0.orig/scripts/generator/cpp/utils.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/cpp/utils.py	2016-09-27 16:01:35.696049479 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2007 Neal Norwitz
 # Portions Copyright 2007 Google Inc.
diff -ur gmock-1.6.0.orig/scripts/generator/gmock_gen.py gmock-1.6.0/scripts/generator/gmock_gen.py
--- gmock-1.6.0.orig/scripts/generator/gmock_gen.py	2011-04-15 13:51:14.000000000 -0600
+++ gmock-1.6.0/scripts/generator/gmock_gen.py	2016-09-27 16:01:27.841974498 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008 Google Inc. All Rights Reserved.
 #
diff -ur gmock-1.6.0.orig/test/gmock_leak_test.py gmock-1.6.0/test/gmock_leak_test.py
--- gmock-1.6.0.orig/test/gmock_leak_test.py	2011-04-15 13:51:15.000000000 -0600
+++ gmock-1.6.0/test/gmock_leak_test.py	2016-09-27 16:02:23.087502544 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2009, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/test/gmock_output_test.py gmock-1.6.0/test/gmock_output_test.py
--- gmock-1.6.0.orig/test/gmock_output_test.py	2011-04-15 13:51:15.000000000 -0600
+++ gmock-1.6.0/test/gmock_output_test.py	2016-09-27 16:02:19.023463622 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2008, Google Inc.
 # All rights reserved.
diff -ur gmock-1.6.0.orig/test/gmock_test_utils.py gmock-1.6.0/test/gmock_test_utils.py
--- gmock-1.6.0.orig/test/gmock_test_utils.py	2011-04-15 13:51:15.000000000 -0600
+++ gmock-1.6.0/test/gmock_test_utils.py	2016-09-27 16:02:20.463477422 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 #
 # Copyright 2006, Google Inc.
 # All rights reserved.
