--- a/setup.py	2016-05-15 19:37:03.101011798 -0400
+++ b/setup.py	2016-05-15 19:37:32.627980976 -0400
@@ -74,7 +74,7 @@
     url='https://github.com/noirbizarre/flask-restplus',
     author='Axel Haustant',
     author_email='axel@data.gouv.fr',
-    packages=find_packages(),
+    packages=find_packages(exclude=['tests', 'tests.*']),
     include_package_data=True,
     install_requires=install_requires,
     tests_require=tests_require,
